### ISSUE TITLE
Increase mutability safety by cloning objects

### ISSUE THREAD

**Issue by KacperKluka** [ably/ably-java#771]
Due to our current API design, it is possible to create a configuration object (e.g. `ClientOptions`), pass it to the SDK and modify it later. This is not ideal as those configuration changes while the SDK is running might lead to unexpected behaviours. The same issue is present when the SDK returns a reference to an object which is used internally and the returned reference allows for modifications (e.g. `LocalDevice`).

To mitigate those issues, we can create and use clones (copies) of such troublesome objects. This will mean that they will become immutable by SDK users after a certain execution point has been reached. For example, we might clone the `ClientOptions` object once it is passed to the `AblyRest` constructor and use only the cloned object. In this case, any further changes to the original object won't affect our SDK configuration. The same goes for "output" objects, instead of returning the original object we should create a copy of it and return it instead. Then, even if a user modifies the copy, the original object will be protected and nothing bad will happen.

The aim of this issue is to find all places in which this cloning can improve safety and implement it there. We should be careful and see if we need to do deep copies (e.g. if an object holds a reference to another object).

┆Issue is synchronized with this [Jira Story](https://ably.atlassian.net/browse/SDK-1861) by [Unito](https://www.unito.io)

**Comment by qsdigor** [comment]
Do we need tests to test the mutability of let's say ClientOptions?

**Comment by ikbalkaya** [comment]
@KacperKluka I just had a call with @qsdigor. If we want to make an object immutable, isn't the best way to mark final all instances down to object reference hierarchy? Currently all non-final and public fields are exposed and some customers might be using them (even if not intended). Using a copy internally might be misleading if you expose it directly to customers. I would favour locking the object from start (or from a point in time) using the structure rather than doing a workaround such as a copy. WDYT?

**Comment by qsdigor** [comment]
If we make object variables final we need to agree on should we lock all of them or maybe leave something to be changed, one variable that sounds changeable is  `logLevel`.

**Comment by qsdigor** [comment]
@KacperKluka You are correct but if we make fields final and change the constructor to implement all fields this should be solved. 
@ikbalkaya Maybe we should consider both approaches as clientId might be mandatory with a few others and the ones that can be changed change but only explicitly in AblyRealtime instance as this `ablyRealtime.options.echoMessages = false`. We can do that by coping object.

**Comment by ikbalkaya** [comment]
> I think that using the final keyword simply means that the object cannot be reassigned but it doesn't mean that you can't modify it, right?
If you mark references `final`  down to the hierarchy it is effectively immutable.

**Comment by KacperKluka** [comment]
You are both correct, but the problem here is that we should not change the public API in the 2.0.0 release :wink: And if we were to use those `final` fields then we would have to change the API and force users to use object constructors (which is not bad itself). Additionally, objects such as `ClientOptions` that have 20+ fields would have enormous constructors and we would probably have to create builders for them (more API changes). So while it is a good idea on its own it doesn't fit our needs and limitations :disappointed:

**Comment by qsdigor** [comment]
We still need to discuss if there are some other objects that would benefit the object cloning/copying and as such, I am reopening the issue.
